### PR TITLE
chisq of sine gaussians

### PR DESCRIPTION
**Pull request by ahnitz**
providing code for reference. Have a look and comment.

### CODE REVIEW

File: bin/pycbc_inspiral [hunk 1]
```diff
@@ -92,10 +92,13 @@ parser.add_argument("--cluster-function", choices=["findchirp", "symmetric"],
 parser.add_argument("--cluster-window", type=float, default = -1,
                     help="Length of clustering window in seconds."
                     " Set to 0 to disable clustering.")
-parser.add_argument("--maximization-interval", type=float, default=0,
+parser.add_argument("--maximization-interval", nargs='+', type=float, default=0,
```
Comment (spxiwh) [inline bin/pycbc_inspiral]: Why this change? (If the answer is not "This is a mistake" the help text would need updating)

File: bin/pycbc_inspiral [hunk 2]
```diff
@@ -92,10 +92,13 @@ parser.add_argument("--cluster-function", choices=["findchirp", "symmetric"],
...
                     help="Maximize triggers over the template bank (ms)")
 parser.add_argument("--bank-veto-bank-file", type=str, help="FIXME: ADD")
-parser.add_argument("--chisq-snr-threshold", type=float, help="Minimum SNR to calculate the power chisq")
+parser.add_argument("--lat-snr-threshold", type=float)
+parser.add_argument("--lat-chisq-locations", type=str, nargs='+')
```
Comment (spxiwh) [inline bin/pycbc_inspiral]: Would need help strings ... But would prefer these options *not* to be in pycbc_inspiral.

Comment (spxiwh) [inline bin/pycbc_inspiral]: (ie. some sort of chi-squared option group for this)

File: pycbc/events/events.py [hunk 3]
```diff
@@ -193,6 +193,20 @@ def newsnr(snr, reduced_x2, q=6., n=2.):
         return newsnr
     else:
         return newsnr[0]
+       
+def bluesnr(snr, bchisq, lchisq):
```
Comment (spxiwh) [inline pycbc/events/events.py]: bluesnr ???

File: pycbc/events/events.py [hunk 4]
```diff
@@ -193,6 +193,20 @@ def newsnr(snr, reduced_x2, q=6., n=2.):
...
+    """ Combined SNR derived from NewSNR and Lat Chisq"""
```
Comment (spxiwh) [inline pycbc/events/events.py]: Should bluesnr be implemented as a new detection statistic, or should it be implemented as a cut (ie. something run at the end of pycbc_inspiral, controlled with a `--lat-chisq-cut = ??` option. That would not then tie the lat-chisq cut to newsnr.

File: pycbc/vetoes/chisq.py [hunk 5]
```diff
@@ -383,11 +383,103 @@ def values(self, corr, snrv, snr_norm, psd, indices, template):
                     rchisq[above] = chisq
             else:
                 rchisq = chisq
-
-            return rchisq, numpy.repeat(dof, len(indices))# dof * numpy.ones_like(indices)
+            dof = numpy.repeat(dof, len(indices))
+            return rchisq, dof
         else:
             return None, None
```
Comment (spxiwh) [inline pycbc/vetoes/chisq.py]: It seems inconsistent to add this in the chisq.py file. In the past we've used one file per type of chi-squared (bank_chisq.py, cont_chisq.py) ... As this one is probably more different than those two, it feels weird to lump it in with Bruce chi-squared.

File: pycbc/waveform/sinegauss.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
+""" Generation of sine-Gaussian bursty type things
```
Comment (spxiwh) [inline pycbc/waveform/sinegauss.py]: ...

**Comment by tdent** [comment]
Hi Alex, 
this patch is a bit of a monster.  Perhaps it would be better to split it into two or more, then the first one would be the actual calculation of the new quantity whatever we call it (I don't know what 'lat' might be .. would suggest 'hfveto' or similar emphasizing that this concerns higher frequencies). 
Then with this please include in comments, or point to, a specification in words of what the code is calculating, otherwise we can't tell if the code is correctly doing what is intended. Currently there are no meaningful comments that I can see.

File: bin/pycbc_inspiral (hunk 2 above)
Comment (ahnitz) [inline bin/pycbc_inspiral]: I'll move to the class itself to provide.

File: pycbc/events/events.py (hunk 3 above)
Comment (ahnitz) [inline pycbc/events/events.py]: It's a test combined single detector statistic. I have not idea what to call it. Come up with something.

File: pycbc/events/events.py (hunk 4 above)
Comment (ahnitz) [inline pycbc/events/events.py]: I think at least initially it should be. It's a lot safer to do so in the sense that we can more easily reconstruct what happened, i.e. we still have triggers to look at and follow up. It also allows us to more easily test different thresholds post-facto until we are happy to move to production (as fast as possible hopefully). When we are *really* comfortable with it and don't feel the need to every check it then depending on how we find to best incorporated it into a single detector statistic it could be done as a veto. It can't be done at the end of the code thought if that's what you mean. Not without regenerating a lot of intermediate products anyway.

File: pycbc/vetoes/chisq.py (hunk 5 above)
Comment (ahnitz) [inline pycbc/vetoes/chisq.py]: Yep, agreed.

**Comment by ahnitz** [comment]
@Tom I'll split off some of the simpler things and rebase this as those get merged.

File: pycbc/vetoes/sgchisq.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
+from pycbc.waveform.utils import apply_fseries_time_shift
+from pycbc.filter import sigma            
+from pycbc.waveform import sinegauss
+from pycbc.vetoes.chisq import SingleDetPowerChisq
+import numpy
+
+class SingleDetSGChisq(SingleDetPowerChisq):
+    """Class that handles precomputation and memory management for efficiently
+    running the power chisq in a single detector inspiral analysis.
```
Comment (tdent) [inline pycbc/vetoes/sgchisq.py]: rewrite comment to say what the class does

File: pycbc/vetoes/sgchisq.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    """
+    def __init__(self, bank, num_bins=0,
+                       snr_threshold=None,
+                       chisq_locations=None):
+        if snr_threshold:
+            self.do = True
+            self.num_bins = num_bins
```
Comment (tdent) [inline pycbc/vetoes/sgchisq.py]: what is num_bins?  what is chisq_locations?  please add comments to explain

File: pycbc/vetoes/sgchisq.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+            Chisq values, one for each sample index
+        """
+        if template.params.template_hash not in self.params:
+            return numpy.ones(len(snrv))
+        
+        values = self.params[template.params.template_hash].split(',')
+        if not self.do:
+            return None
+        
+        # Get the chisq bins to use as the frequency reference point
```
Comment (tdent) [inline pycbc/vetoes/sgchisq.py]: not clear.  comment could say something about why we need a frequency reference point

File: pycbc/vetoes/sgchisq.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+        bins = self.cached_chisq_bins(template, psd)
+
+        # This is implemented slowly, so let's not call it often, OK?
+        chisq = numpy.ones(len(snrv))
+        for i in range(len(snrv)):
+            snr = abs(snrv[i] * snr_norm)
+            if snr < self.snr_threshold:
+                continue
+
+            N = (len(template) - 1) * 2
```
Comment (tdent) [inline pycbc/vetoes/sgchisq.py]: i dont know what is being done here

File: pycbc/vetoes/sgchisq.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+            dt = 1.0 / (N * template.delta_f)
+            kmin = int(template.f_lower / psd.delta_f)
+            time = float(template.epoch) + dt * indices[i]
+            # Shift the time of interest to be centered on 0
+            stilde_shift = apply_fseries_time_shift(stilde, -time)
+
+            qwindow = 50
+            chisq[i] = 0
+            fpeak = (bins[-2] * 2.0 - bins[-3]) * template.delta_f
+            fstop = len(template) * template.delta_f * 0.9
```
Comment (tdent) [inline pycbc/vetoes/sgchisq.py]: i dont know what is being done here either.  indices -2 and -3 seem hard coded without further explanation

File: bin/hdfcoinc/pycbc_fit_sngls_binned [hunk 12]
```diff
@@ -35,14 +35,17 @@ stat_dict = {
...
     "snr"           : lambda snr, rchisq : snr,
-    "snronchi"      : lambda snr, rchisq : snr / (rchisq ** 0.5)
+    "snronchi"      : lambda snr, rchisq : snr / (rchisq ** 0.5),
+    "blue_snr"      : events.bluesnr,
 }
 
-def get_stat(statchoice, snr, rchisq, fac):
+def get_stat(statchoice, snr, rchisq, sgchisq, fac):
     if fac is not None:
         if statchoice not in ["new_snr", "effective_snr"]:
```
Comment (tdent) [inline bin/hdfcoinc/pycbc_fit_sngls_binned]: add the new stat name to this input checking list

File: bin/hdfcoinc/pycbc_fit_sngls_binned [hunk 13]
```diff
@@ -35,14 +35,17 @@ stat_dict = {
     "new_snr"       : events.newsnr,
     "effective_snr" : events.effsnr,
     "snr"           : lambda snr, rchisq : snr,
-    "snronchi"      : lambda snr, rchisq : snr / (rchisq ** 0.5)
+    "snronchi"      : lambda snr, rchisq : snr / (rchisq ** 0.5),
+    "blue_snr"      : events.bluesnr,
```
Comment (tdent) [inline bin/hdfcoinc/pycbc_fit_sngls_binned]: i dont think we want to have to answer to 50 people as to why 'blue snr' makes sense. prefer to start with something rather more wonky like 'newsnr_sgveto' ?

File: bin/hdfcoinc/pycbc_fit_sngls_binned [hunk 14]
```diff
@@ -176,12 +179,19 @@ snr = snr[aboveminth]
 chisq = trigf[args.ifo+'/chisq'][:][aboveminth]
 chisq_dof = trigf[args.ifo+'/chisq_dof'][:][aboveminth]
 rchisq = chisq / (2 * chisq_dof - 2)
+
+try:
+    sgchisq = trigf[args.ifo+'/sg_chisq'][:][aboveminth]
+except:
```
Comment (tdent) [inline bin/hdfcoinc/pycbc_fit_sngls_binned]: prefer to have 'except KeyError' or whatever will happen if you try this on a file with no sgchisq value.

File: bin/hdfcoinc/pycbc_fit_sngls_by_template [hunk 15]
```diff
@@ -137,14 +139,19 @@ templatef = h5py.File(args.bank_file, 'r')
 chisq = trigf[args.ifo+'/chisq'][:]
 chisq_dof = trigf[args.ifo+'/chisq_dof'][:]
 rchisq = chisq / (2 * chisq_dof - 2)
+try:
+    sgchisq = trigf[args.ifo+'/sg_chisq'][:]
+except:
+    sgchisq = None
```
Comment (tdent) [inline bin/hdfcoinc/pycbc_fit_sngls_by_template]: what will happen if you feed None to the function that calculates the new statistic?  do we prefer to assign it to an array of zeros in case the file doesn't contain any values?

File: pycbc/events/stat.py [hunk 16]
```diff
@@ -48,6 +48,25 @@ def get_newsnr(trigs):
     newsnr = events.newsnr(trigs['snr'], trigs['chisq'] / dof)
     return numpy.array(newsnr, ndmin=1, dtype=numpy.float32)
 
+def get_bluesnr(trigs):
+    """
+    Calculate bluesnr ('reweighted SNR') for a trigs object
```
Comment (tdent) [inline pycbc/events/stat.py]: remove the explanation in brackets as it isn't true any more

File: pycbc/events/stat.py [hunk 17]
```diff
@@ -413,6 +434,16 @@ def coinc(self, s0, s1, slide, step):
         # scale to resemble network SNR
         return cstat / (2.**0.5)
 
+class PhaseTDExpFitMultiStatistic(PhaseTDExpFitStatistic):
```
Comment (tdent) [inline pycbc/events/stat.py]: hmm. don't know where 'Multi' fits in here .. maybe it would be better to rename more stuff to clear up the distinction between the newsnr-based combined stat and the newsnr+sg one 
eg phasetd_exp_fit_stat -> phasetd_exp_fit_newsnr 
then new thing is phasetd_expfit_newsnr_sg[veto]

File: bin/hdfcoinc/pycbc_fit_sngls_by_template (hunk 15 above)
Comment (ahnitz) [inline bin/hdfcoinc/pycbc_fit_sngls_by_template]: It will blow up at you. I think that's actually the right behavior since it will tell you it is missing this information, and then you can figure out why.

File: pycbc/events/stat.py (hunk 17 above)
Comment (ahnitz) [inline pycbc/events/stat.py]: I'd rather leave the original name the same.

File: bin/hdfcoinc/pycbc_fit_sngls_binned [hunk 18]
```diff
@@ -35,14 +35,18 @@ stat_dict = {
...
     "snr"           : lambda snr, rchisq : snr,
-    "snronchi"      : lambda snr, rchisq : snr / (rchisq ** 0.5)
+    "snronchi"      : lambda snr, rchisq : snr / (rchisq ** 0.5),
+    "newsnr_sgveto"      : events.newsnr_sgveto,
 }
 
-def get_stat(statchoice, snr, rchisq, fac):
+def get_stat(statchoice, snr, rchisq, sgchisq, fac):
+    if statchoice not in stat_dict:
+        raise RuntimeError("Can't use --stat-factor with this statistic!")
```
Comment (tdent) [inline bin/hdfcoinc/pycbc_fit_sngls_binned]: this error message now makes no sense.  if statchoice is not one of the allowed statistics, the problem is nothing to do with the stat-factor argument

File: bin/hdfcoinc/pycbc_plot_singles_vs_params [hunk 19]
```diff
@@ -121,9 +121,14 @@ if opts.z_var == 'density':
     norm = LogNorm()
     hb = ax.hexbin(x, y, norm=norm, vmin=1, **hexbin_style)
     fig.colorbar(hb, ticks=LogLocator(subs=range(10)))
-elif opts.z_var in ['max(snr)', 'max(newsnr)']:
+elif opts.z_var in ['max(snr)', 'max(newsnr)', 'max(newsnr_sgveeto)']:
```
Comment (tdent) [inline bin/hdfcoinc/pycbc_plot_singles_vs_params]: typo!

**Comment by tdent** [comment]
Well, I am using the word 'veto' because the sg_chisq value only gets used for a small minority of triggers, most of the triggers are not affected by it.  In reality there is a continuum between DQ-type vetoing where you effectively set the ranking statistic to -inf, and Alex's first idea where you knock it down to a fixed, small value, and the current statistic ...
